### ISSUE TITLE
VS 2019 16.11.0 Preview 2.0 - Crash because Context is null in ArrayAdapter constructor

### ISSUE THREAD

**Issue by gmck** [xamarin/xamarin-android#6152]
I’m getting the following crash listed at the end of this issue in my SubscriptionFragment because the Activity (context) passed in the OnResume() of the fragment to the ArrayAdapter is null in the code listed below.

This only occurs if I force the fragment to display in Portrait if the app is already in Landscape when the user selects the purchase subscription menu item. If the menu item is selected when the screen is already in Portrait mode, then there is no problem.

The OnResume() calls both QuerySkuDetailsAsync and QueryPurchaseHistoryAsync to populate an` IList<SkuDetails>` and an `IList<PurchaseHistoryRecord>` respectively. When debugging the crash with a breakpoint before the ArrayAdapter’s ctor the SkuItems.Count is > 0 and the PurchaseHistoryAdapter is already correctly populated.

I originally had the OnResume() code in the OnStart() but moved it to OnResume hoping the context would be at least viable by then. If it is not viable by OnResume then I guess it will never be!!!

The reason I want the SubscriptionFragment to be Portrait only is that the design of the content of the layout is too large to comfortably fit in Landscape mode on a phone, without resorting to wrapping the androidx.constraintlayout.widget.ConstraintLayout in an androidx.core.widget.NestedScrollView. 

I tried it and I just don’t like scrolling such a small area for what I consider is such an important fragment of the app.

I’m using QueryPurchaseHistoryAsync because it always returns a value as compared to QueryPurchases which doesn’t if the purchase has expired or is on hold or paused. Additionally, the click item event of the PurchaseHistoryAdapter allows me to get the PurchaseToken and the Sku of the selected item which I can then use to immediately validate the subscription on the server and return a Google.Apis.AndroidPublisher.v3.Data.SubscriptionPurchase which allows me to open another fragment displaying the values of the returned SubscriptionPurchase. Those details are really valuable for debugging the torturous route of testing integration of Xamarin.Android.Google.BillingClient v3 or v4 in the 5, 15 etc minute intervals you get, to see if your subscription is reacting correctly at each step. I suspect some of that info could be also informational to users if they have a requirement to check the status of their subscription within the app.

I can get around the crash by using OnAttach and saving the context passed to a local member of the fragment and then pass that to the ArrayAdapter and it no longer crashes. I can also substitute Android.App.Application.Context for Activity and that also prevents the crash, but I'm not sure that is a really valid use of Application.Context.

However even though I have a workaround, for now, I still think I should report this. It seems from the crash report from App Center that it is related to System.Runtime.CompilerServices.AsyncMethodBuilderCore.

```
public override async void OnResume()
        {
            base.OnResume();

            List<string> skus = new List<string> { GetString(Resource.String.sku_twelve_months), GetString(Resource.String.sku_six_months) };

            SkuDetailsParams skuDetailsParams = SkuDetailsParams.NewBuilder()
                                                               .SetSkusList(skus)
                                                               .SetType(BillingClient.SkuType.Subs)
                                                               .Build();

            // Two native Queryxx..Async methods. So we use await inAppBilling.ConnectAsync.
            // Note Billling 4.0.0 now has QueryPurchasesAsync as compared to just QueryPurchases in 3.0.0 so we will need to change it to when we upgrade
            if (await inAppBilling.ConnectAsync())
            {
                QuerySkuDetailsResult querySkuDetailsResult = await inAppBilling.BillingClient.QuerySkuDetailsAsync(skuDetailsParams);

                if (querySkuDetailsResult.Result.ResponseCode == BillingResponseCode.Ok && querySkuDetailsResult.SkuDetails != null)
                {
                    skuDetails = querySkuDetailsResult.SkuDetails;
                    skuItems = new List<string>();
                    for (int i = 0; i < skuDetails.Count; i++)
                    {
                        skuDetail = new SkuDetails(skuDetails[i].OriginalJson);
                        if (skuDetail != null)
                        {
                            string title = skuDetail.Title.Contains("(") ? skuDetail.Title.Substring(0, skuDetail.Title.IndexOf("(")) : skuDetail.Title;
                            string description = skuDetail.Description;
                            string price = skuDetail.Price;
                            string currency = skuDetail.PriceCurrencyCode;
                            string item = title + " " + description + " " + price;
                            skuItems.Add(item);
                        }
                    }
                }
                else if (querySkuDetailsResult.Result.ResponseCode == BillingResponseCode.BillingUnavailable)
                    BillingProcess.ShowGooglePlayUnavailableDialog(Activity, GetString(Resource.String.google_play_billing_unavailable_title), GetString(Resource.String.google_play_billing_unavailable_explanation));


                // Ctor for PurchaseHistoryRecord but not using  - purchaseHistoryRecord = new PurchaseHistoryRecord(purchaseHistoryRecords[0].OriginalJson, purchaseHistoryRecords[0].Signature);  
                QueryPurchaseHistoryResult queryPurchaseHistoryResult = await inAppBilling.BillingClient.QueryPurchaseHistoryAsync(BillingClient.SkuType.Subs);
                if (queryPurchaseHistoryResult.Result.ResponseCode == BillingResponseCode.Ok)
                {
                    if (queryPurchaseHistoryResult.PurchaseHistoryRecords != null)
                    {
                        purchaseHistoryRecords = queryPurchaseHistoryResult.PurchaseHistoryRecords;
                        purchaseHistoryRecords.OrderByDescending(p => p.PurchaseTime);
                    }
                }
                else if (queryPurchaseHistoryResult.Result.ResponseCode != BillingResponseCode.Ok)
                {
                    // Flesh this out
                }
                inAppBilling.Disconnect();
            }

            // Changed the order here (was the skus then the purchaseHistory). It didn't make any difference to the crash, but it appears to lessen the delay in populating the history records.
            
// Populate the purchaseHistoryAdapter
            if (purchaseHistoryRecords != null && purchaseHistoryRecords.Count > 0)
            {
                purchaseHistoryAdapter.Clear();
                foreach (PurchaseHistoryRecord purchaseHistoryRecord in purchaseHistoryRecords)
                    purchaseHistoryAdapter.AddItem(purchaseHistoryRecord);
            }

            // Populate the skuArrayAdapter and enable the purchase button
            if (skuItems != null && skuItems.Count > 0)
            {
                // See OnAttach for this context. What a drama... Update - still another crash if context made null in OnDestroy. Get a heap of calls between MainActivity's OnDestroy and then OnAttach in the
                // debugger. Obviously some problem with the debugger because if you log it then there in only two calls to OnAttach.
                // Android.App.Application.Context works, so why not use that and be done with it.??
                ArrayAdapter skuArrayAdapter = new ArrayAdapter(Activity, /*context,*/ /*Android.App.Application.Context,*/ Resource.Layout.autocomplete_textview_item, skuItems.ToArray());
                autoCompleteTextView.Adapter = skuArrayAdapter;
                purchaseButton.Enabled = true;
            }
            else
                autoCompleteTextView.Text = GetString(Resource.String.purchase_no_product_available); 
        }
```
Sorry about the above code formatting - but that is how it worked out!!.
 
Crash details from App  Center

Package: com.glmsoftware.OBDNowPro
Version Code: 157
Version Name: 1.57
Android: 11
Android Build: RQ3A.210705.001
Manufacturer: Google
Model: Pixel 4a
CrashReporter Key: 4292e1f2-c73a-446f-9162-884e8c4107e9
Start Date: 2021-07-30T01:00:32.827Z
Date: 2021-07-30T01:01:35.479Z

Xamarin Exception Stack:
Java.Lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
  at Java.Interop.JniEnvironment+InstanceMethods.CallNonvirtualVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x0008e] in <b3d3e7a78651409b978d7cb682dcc5ca>:0
  at Java.Interop.JniPeerMembers+JniInstanceMethods.FinishCreateInstance (System.String constructorSignature, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x0004f] in <b3d3e7a78651409b978d7cb682dcc5ca>:0
  at Android.Widget.ArrayAdapter..ctor (Android.Content.Context context, System.Int32 resource, System.Collections.IList objects) [0x000aa] in <59d26097ee3540c3ba985d9e2a57150a>:0
  at com.glmsoftware.OBDNowPro.Fragments.SubscriptionFragment.OnResume () [0x00613] in <75217d8466c544228534f4f6cafaa02e>:0
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) [0x00000] in <1b39a03c32ec46258a7821e202e0269f>:0
  at Android.App.SyncContext+<>c__DisplayClass2_0.<Post>b__0 () [0x00000] in <59d26097ee3540c3ba985d9e2a57150a>:0
  at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <59d26097ee3540c3ba985d9e2a57150a>:0
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <59d26097ee3540c3ba985d9e2a57150a>:0
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.35(intptr,intptr)
  at java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
  at android.view.LayoutInflater.from(LayoutInflater.java:282)
  at android.widget.ArrayAdapter.<init>(ArrayAdapter.java:213)
  at android.widget.ArrayAdapter.<init>(ArrayAdapter.java:207)
  at android.widget.ArrayAdapter.<init>(ArrayAdapter.java:193)
  at crc64f0146600faa7a777.InternalPurchaseHistoryResponseListener.n_onPurchaseHistoryResponse(Native Method)
  at crc64f0146600faa7a777.InternalPurchaseHistoryResponseListener.onPurchaseHistoryResponse(InternalPurchaseHistoryResponseListener.java:30)
  at com.android.billingclient.api.zzm.run(com.android.billingclient:billing@@3.0.0:4)
  at android.os.Handler.handleCallback(Handler.java:938)
  at android.os.Handler.dispatchMessage(Handler.java:99)
  at android.os.Looper.loop(Looper.java:223)
  at android.app.ActivityThread.main(ActivityThread.java:7664)
  at java.lang.reflect.Method.invoke(Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

**Comment by grendello** [comment]
@gmck it looks like the crash originates in Java when inflating a layout, not sure how Xamarin.Android is involved here. @jpobst, any idea?

**Comment by gmck** [comment]
@jpobst 
Thanks for that link. Over the weekend I moved the code from OnResume back into OnStart. Then I realised I was initialising the  PurchaseHistoryAdapter in OnViewCreated, so why not do the same for the ArrayAdapter. So once I did that it was only a matter of updating the ArrayAdapter in OnStart with skuArrayAdapter.AddAll(skuItems);

So I now don't need the context from OnAttach and the other workarounds I tried.

I'm pretty sure that is the first time I've ever used an ArrayAdapter in all the years I've been using Xamarin. After this episode, I doubt I'll forget the correct way to use it.

It's kind of ironic, but in the OnResume there is a call to Show(Activity.SupportFragmentManager, tag) to call a BottomSheetDialogFragment explaining to first-time users that they can click on a purchaseHistoryRecord to get more info on the subscription status. No problems about using Activity there.

Makes you wonder if C# shouldn't have kept with GetActivity() and GetContext() etc rather than the getters. 

That guys getContextNullSafety() in that thread was pretty full-on!!!.
